pkg/provider/azure_controller_common.go

```diff
@@ -442,70 +442,67 @@ func (c *controllerCommon) GetDiskLun(diskName, diskURI string, nodeName types.N
 	return -1, provisioningState, fmt.Errorf("%s for disk %s", consts.CannotFindDiskLUN, diskName)
 }
 
-// SetDiskLun find unused luns and allocate lun for every disk in diskMap.
-// Return lun of diskURI, -1 if all luns are used.
-func (c *controllerCommon) SetDiskLun(nodeName types.NodeName, diskURI string, diskMap map[string]*AttachDiskOptions) (int32, error) {
+// SetDiskLun find unused luns and allocate lun for every disk in disksPendingAttach map.
+// Return err if not enough luns are found.
+func (c *controllerCommon) SetDiskLun(nodeName types.NodeName, disksPendingAttach map[string]*AttachDiskOptions) error {
 	disks, _, err := c.getNodeDataDisks(nodeName, azcache.CacheReadTypeDefault)
 	if err != nil {
 		klog.Errorf("error of getting data disks for node %s: %v", nodeName, err)
-		return -1, err
+		return err
 	}
 
-	lun := int32(-1)
-	_, isDiskInMap := diskMap[diskURI]
-	used := make([]bool, maxLUN)
+	allLuns := make([]bool, maxLUN)
+	uriToLun := make(map[string]int32, len(disks))
 	for _, disk := range disks {
 		if disk.Lun != nil {
-			used[*disk.Lun] = true
-			if !isDiskInMap {
-				// find lun of diskURI since diskURI is not in diskMap
-				if disk.ManagedDisk != nil && strings.EqualFold(*disk.ManagedDisk.ID, diskURI) {
-					lun = *disk.Lun
-				}
+			allLuns[*disk.Lun] = true
```

Comment:
We should make sure that `*disk.Lun` is in the range [0..64], otherwise this will panic.